### ISSUE TITLE
Allow Proposers to Attach Additional Metadata to Proposals

### ISSUE THREAD

**Issue by davidterpay** [cometbft/cometbft#1670]
### Summary

Allow proposers to attach arbitrary data to proposals built with ABCI++ (i.e. `PrepareProposal`, `ProcessProposal`) without having to inject it into the proposal itself.

### Problem Definition

At Skip we utilize ABCI++ for our in-protocol oracle (Slinky) and block builder (Block SDK). We attach related price updates and info about the block by inserting this information into the first slot of the proposal. This is a bit annoying for chains since it is expected that the first `Tx` always fails. Proposers should ideally be able to attach arbitrary information to proposals which other network operators can utilize as the protocol defines without inhibiting the process of executing transactions in a block.

This is an issue today because `ResponsePrepareProposal` only returns a list of Txs.

```golang
type ResponsePrepareProposal struct {
	Txs [][]byte `protobuf:"bytes,1,rep,name=txs,proto3" json:"txs,omitempty"`
}
```

### Proposal

One simple approach to solving this would be to add an additional field - `Info` - to be added to the following response/request types:
- `ResponsePrepareProposal`
- `RequestProcessProposal`
- `FinalizeBlock`

#### ResponsePrepareProposal

This would allow the proposer to inject arbitrary data about the block, the vote extensions from the previous block, and more as they are constructing a block. Providers a clean abstraction for work to be done with vote extensions and block building.

#### RequestProcessProposal

This would allow the network operators to verify additional data that should be included in consensus and/or may be committed to state. 

#### RequestFinalizeBlock

This would allow data included in the new `Info` field to be written to state.

**Comment by melekes** [comment]
Is this a CometBFT or Cosmos SDK issue? 

Quoting Zaki: "For Comet, txs are just data blobs, and the Cosmos SDK just needs to know that a data blob isn't a tx, which Skip is jury-rigging.". Quoting Bez: "In v0.50 Cosmos SDK we handle this".

So, if Cosmos SDK (> v0.50) handles this already, should we still add the Info field?

**Comment by davidterpay** [comment]
The TLDR on why we believe this is a good add is because the proposer's should be able to relay information about block proposals to other validators in the network without having to include that additional information in the block proposal - which may end up being included in history of the chain, may likely have unintended consequences on applications running on comet, and provide a UX/UI barrier (mostly specific to cosmos).

For example, a proposer might set some additional conditions that they want other validator's in the network to check that is independent of the block proposal that was constructed. These could be invariants, vote extensions, or other.

The fix that `v0.50.0` has is a relatively hacky solution to this problem, adding this additional field is the best solution from our perspective.

**Comment by cason** [comment]
The suggestion is therefore to include an additional field to the Block which is not written to the blockchain, being exclusively pertinent to the exchange of the proposed blocks between nodes?

I assume that this additional `Info` field must be signed. But its signature should not be the same as the proposed Block, since this field is not persisted (to the blockchain and other stores). It also would not be retrieved via block sync, I would assume, so when delivering such blocks to the application, as part of the initialization via block sync, this field will be missing, right? This entails some problems in terms of the state of the application, which, as you mentioned, is also updated based on the information provided by the `Info` field.

In summary, it is not a trivial change, it breaks the block format or the consensus messages format, and have potential impact on the recovery of a crash node or the bootstrap of a new node.

**Comment by sergio-mena** [comment]
@cason 

> The suggestion is therefore to include an additional field to the Block which is not written to the blockchain, being exclusively pertinent to the exchange of the proposed blocks between nodes?

My understanding is that we add a field (or several) to the proposal _message_, not to the proposal itself.

About the need for signing, I guess it'll depend on the use case (e.g., if you just want to send the vote extensions received, you don't need to sign anything). This will need to be discussed when we tackle this as part of the ABCI 3.0 effort.

Agree the change is not trivial. I added this issue to #9, to be done with other ABCI improvements when we include it as part of our planning. Will be similar to how we came up with ABCI 2.0 (probably bi-weekly meetings with a quorum of prominent stakeholders).

**Comment by itsdevbear** [comment]
+1 here

**Comment by cason** [comment]
> My understanding is that we add a field (or several) to the proposal message, not to the proposal itself.

The consensus' `Proposal` message? The assumption we have is that it has a fixed-size.

**Comment by sergio-mena** [comment]
©cason It's not clear to me at which level this data should be inserted, we will need to look into it.

In fact, by quickly looking at the (current) code, this is what seems clear to me at this point:

* `State`'s method `createProposalBlock` should return a block *and the extra data*. This extra data should be obtained from the app via extending `PrepareProposalResponse` with a new field
* This new data item (probably a `byte[]` at CometBFT level) should be added to the proposal message
  * either inside `types.Proposal`
  * or outside `types.Proposal`, directly added to `ProposalMessage`

I think it's too early to say where it will go. The need for this to be signed, or not (some work looking at use cases needed here) will have an impact. Also, we should try to avoid breaking the way blocks are verified (which is considered part of the "block format"), so it is unlikely we will be modifying `CanonicalProposal`. Rather, we'll likely add another canonical structure for this data. Similarly to what we did with vote extensions, which are not part of `CanonicalVote` but signed independently to avoid breaking the way vote signatures are verified and thus breaking the block format.

Another thought is that we might have similar problems to those with vote extension sizes. We might take the opportunity and try to come with a holistic solution for both.

In short, we will tackle this when we make #9 one of our top prios in the coming months.

**Comment by cason** [comment]
Ok, for me the `Proposal` message is a consensus-level message that carries the reference to a propose block, lets say a payload. The `Proposal` by itself has no use, it is meant to know the payload we are looking for. In fact, the block is not signed when shipped using `Part` messages. The `Proposal` is signed and we check whether the payload matches the proposed one (`BlockID` has the merkle root for the payload/block and its has).

If we plan to add any new content, which is a payload, namely something with a variable size, we should probably include it in the block _or_ ship it together with the block (using `Part` messages). Not in the `Proposal`. Notice that in PBTS we use the timestamp of the `Proposal` to evaluate whether the block timestamp is a good one. This is because the block can take much longer than the proposal to be delivered. Because the block can be pretty large and the Proposal has a fixed size.
